### ISSUE TITLE
Becomes Noisy after xrun

### ISSUE THREAD

**Issue by Rippert** [brummer10/Rumor#2]
I've noticed that the plugin "switches" from a low noise, full bass sound, to a high noise trebly sound after an xrun occurs sometimes. This seems to be a permanent change as long as the plugin is still running. turning it off and on, such as with the "power" button in the Guitarix rack will reset it to normal. In `jalv`, I need to destroy the instance and reinitialize it to get it to reset.

**Comment by brummer10** [comment]
Which sample rate / frame size do you use? I can't reproduce the "high noise trebly sound after an xrun". But I've implemented a "clear state" func now, when the plug is bypassed. So, when that happen again, in jalv, and you can't reset it to normal by bypass it, there must be a other source for the issue then the internal filter (the resampler comes in mind then).

**Comment by Rippert** [comment]
I was using 96000 Hz at 128 samples yesterday. I switched to 64 samples today to get the instability to show up faster. Still, it takes a while before it happens, and I had to open up the Convolver in Guitarix to increase the xruns and play for awhile before it would happen. Sounds like some kind of bi-stable filter. I can see in the code that you are using an IIR filter, maybe it has an instability? It's not easy to make switch, so it may be hard to pin down.

**Comment by Rippert** [comment]
A little Off topic, but I noticed that the  rumor.dsp file mentions that Rumor is based on the Ruiner circuit, but you also have a Ruiner plugin built into guitarix, and it isn't anything like the Rumor plugin. Or, at least, the Runier plugin is much lower gain for some reason. Are they supposed to both be based on the same circuit?

**Comment by brummer10** [comment]
So I've switched the filter type to a direct form Biquad Filter. That should stabilise the Filter.

Yes, it's based on the same circuit, just that with the Rumor I do a new interpretation for the non-linear part.
I think this one cover the behave of the original a bit better.

**Comment by Rippert** [comment]
New version still changes after a lot of xruns, but isn't noisy, just trebly. I turned off all my CPU affinity stuff, and couldn't reproduce it, so maybe too much of a niche case to keep pursuing.

**Comment by brummer10** [comment]
The question now is, did it switch back to normal when you switch it off/on (bypass). If not, then it isn't the filter which explode, but some side stuff.

**Comment by Rippert** [comment]
I did a checkout of  https://github.com/brummer10/Rumor/commit/27f1626899de6fb57caa158a450d5e4c1c70f224 and recompiled. Then I ran the standalone LV2 in `jalv` overnight in an isolated CPU core, which is what first caused the issue. When I just tested it, there was no sound at all. I hit the bypass switch and could hear the clean bass tone, and then hit the bypass switch again and the normal, distorted tone came on. So not the same as before, but does suggest that the reset from the bypass switch fixes something. I'll leave it again overnight and see what happens, then maybe try it with the latest commit.

Merry Christmas Hermann

**Comment by Rippert** [comment]
The second overnight test went the same as the first, no sound until turning the effect off with the bypass switch, then normal clean sound and normal distorted sound when effect switched back on. I've recompiled with the latest commit, and will try again.

**Comment by brummer10** [comment]
So it's clearly the filter which sokes up. Unfortunately I've no idea why, or how that could happen. I could implement vectorisation for the filter, but I guess, for a isolated CPU core that wouldn't gain any improvement. 
Seems to me like a stack compromising thing which I don't understand.

**Comment by Rippert** [comment]
Tried it with the latest commit, and got the same results. I'll turn off the CPU isolation stuff and try again. Seems like leaving it overnight is the only reliable way to get the failure mode to occur when using the standalone setup with `jalv`. We'll see if that's true with normal CPU usage.

**Comment by Rippert** [comment]
OK, got the same result again without any CPU isolation. I'll try running Guitarix with no plugins overnight and see if it loses sound and needs to be reset just as a control to be sure its something specific to Rumor.

**Comment by brummer10** [comment]
Okay, I guess I've found the culprit. Seems that this filter have a problem with denormals. I've implemented a protector now.

**Comment by brummer10** [comment]
mmm, last commit still break after 10 hours. Just started a new try with a slightly change, will commit it if it hold 10 hours. :)

**Comment by Rippert** [comment]
The test with just Guitarix was still working fine after sitting overnight, as you probably guessed.

I'm glad you were able to reproduce the issue. I always worry that I'm just doing something stupid on my system.

**Comment by brummer10** [comment]
I'm slowly getting somewhere. It seems that the low-pass filter used before downsampling is the culprit.
I've implemented a printout function to check the content of the internal buffers after the plug becomes silent. I found values of 
610055110715593,875000 in the lowpass buffer, were it should be zero. 
I changed the filter order to 4. now and started a new check to see how the buffer beehive. So far it looks good, as the content stay at zero when the plug idle. Now give it some hours. If it works I'll push a update to the repository.
